terminal-emulator/src/main/java/com/termux/terminal/TerminalEmulator.java

```diff
@@ -0,0 +1,2447 @@
+                            if (code == 38) {
+                                mForeColor = argbColor;
+                            } else {
+                                mBackColor = argbColor;
+                            }
+                        }
+                        i += 4; // "2;P_r;P_g;P_r"
+                    }
+                } else if (firstArg == 5) {
+                    int color = mArgs[i + 2];
+                    i += 2; // "5;P_s"
+                    if (color >= 0 && color < TextStyle.NUM_INDEXED_COLORS) {
+                        if (code == 38) {
+                            mForeColor = color;
+                        } else {
+                            mBackColor = color;
+                        }
+                    } else {
+                        if (LOG_ESCAPE_SEQUENCES) mClient.logWarn(LOG_TAG, "Invalid color index: " + color);
+                    }
+                } else {
+                    finishSequenceAndLogError("Invalid ISO-8613-3 SGR first argument: " + firstArg);
+                }
+            } else if (code == 39) { // Set default foreground color.
+                mForeColor = TextStyle.COLOR_INDEX_FOREGROUND;
+            } else if (code >= 40 && code <= 47) { // Set background color.
+                mBackColor = code - 40;
+            } else if (code == 49) { // Set default background color.
+                mBackColor = TextStyle.COLOR_INDEX_BACKGROUND;
+            } else if (code >= 90 && code <= 97) { // Bright foreground colors (aixterm codes).
+                mForeColor = code - 90 + 8;
+            } else if (code >= 100 && code <= 107) { // Bright background color (aixterm codes).
+                mBackColor = code - 100 + 8;
+            } else {
+                if (LOG_ESCAPE_SEQUENCES)
+                    mClient.logWarn(LOG_TAG, String.format("SGR unknown code %d", code));
+            }
+        }
+    }
+
+    private void doOsc(int b) {
+        switch (b) {
+            case 7: // Bell.
+                doOscSetTextParameters("\007");
+                break;
+            case 27: // Escape.
+                continueSequence(ESC_OSC_ESC);
+                break;
+            default:
+                collectOSCArgs(b);
+                break;
+        }
+    }
+
+    private void doOscEsc(int b) {
+        switch (b) {
+            case '\\':
+                doOscSetTextParameters("\033\\");
+                break;
+            default:
+                // The ESC character was not followed by a \, so insert the ESC and
+                // the current character in arg buffer.
+                collectOSCArgs(27);
+                collectOSCArgs(b);
+                continueSequence(ESC_OSC);
+                break;
+        }
+    }
+
+    /** An Operating System Controls (OSC) Set Text Parameters. May come here from BEL or ST. */
+    private void doOscSetTextParameters(String bellOrStringTerminator) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 33. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.